### ISSUE TITLE
Doesn't work with the latest LanguageTool version - 6.0

### ISSUE THREAD

**Issue by Hell01World** [dpelle/vim-LanguageTool#33]
When I type :LanguageToolCheck, and it uses the latest version of LanguageTool - 6.0 it outputs an error message. It's quite long, so I will just [send it here as a text file.](https://github.com/dpelle/vim-LanguageTool/files/10351900/error_message.txt)

Also, I found this in the LanguageTool's CHANGES.md:
> * The `--api` parameter for the command-line version has been removed. It had long been deprecated and replaced by `--json`.

Which is probably the reason why it no longer works because this error starts with:
> `['Unknown parameter: --api'`

It works fine with version 5.9

**Comment by tarikgraba** [comment]
:+1: it seems they abandoned the XML output (it was marked as deprecated) in favor of the sole JSON.

**Comment by dpelle** [comment]
Yes, I am aware of the reason why the Vim plugin does not work latest LanguageTool.
I need to update the plugin to use the JSON format, which in principle should make the plugin simpler, since Vim understands JSON nowadays. However, the JSON response from LanguageTool no longer includes the line/column of each error which is unfortunate. It instead gives the Unicode character offset from the beginning of the file. There is no good function in Vim to convert it in line/column, although there is an "ugly" workaround (not sure how slow it will be).  

More info at:
- https://groups.google.com/g/vim_dev/c/AVpp8DT2_Vc/m/NZ9h3F8gBQAJ
- https://github.com/languagetool-org/languagetool/issues/7462
- https://github.com/languagetool-org/languagetool/issues/991
- workaround at: https://github.com/languagetool-org/languagetool/issues/991#issuecomment-521026834

And unfortunately, I do not have a lot of spare time these days for this, but I have not forgotten.
I was hoping to fix it during my Christmas holidays, but I did not.

**Comment by tobalr** [comment]
Older versions of LanguageTool can be downloaded from their download page https://languagetool.org/download/
Version 5.9 seems to be the latest working version

**Comment by revolutionnaire** [comment]
Where is a good place to unzip the old version to? Homebrew doesn't provide older versions, and it installs it in the `/opt/` folder. Is it better to place it there also?

**Comment by Britaliope** [comment]
@dpelle Any idea if you'll have time to look at this at some point ?

I don't have much time myself, but i'd like to help anyway, however i don't have that much experience with vimscript.

What was the fix you were hoping to work on last christmas : was it the workaround you linked above, or a patch in Languagetool to publish the needed information in the API ?

**Comment by Britaliope** [comment]
For people looking for an alternative tool: I switched to TeXtidote, using https://github.com/PatrBal/vim-textidote which works correctly using latest versions of Languagetool. It also have the added benefits of parsing the latex syntax, so it don't trigger grammar errors on latex keywords or \usepackage names.

**Comment by dpelle** [comment]
I will try to have a look at fixing this (switching to the "new" JSON API instead of the new defunct XML API) in my Chrismas holidays. Unfortunately, the JSON API does not contain the same information as the previous XML API which complicates things.

**Comment by Britaliope** [comment]
> I just glanced at https://github.com/PatrBal/vim-textidote/blob/master/autoload/textidote.vim and it appears to be using XML too so unless I'm missing something, it should also not work with the latest LanguageTool. It in fact mostly copied the parsing code from my plugin.

vim-textidote integrate with textidote (https://github.com/sylvainhalle/textidote), which provide spell/grammar checking on LaTeX and Markdown documents using LanguageTool as a backend. However vim-textidote can also integrate directly with LanguageTool, buypassing textidote, and this can be used for other documents that are not LaTeX or Markdown. This integration is indeed heavily based on yours (as stated in the readme).

The LanguageTool integration is indeed not compatible with newer versions of languagetool. However, the textidote version relies on textidote, which is a java applet that directly uses the languagetool Jar so it doesn't have the xml vs json API problem

**Comment by Toreno96** [comment]
If you're using Homebrew to install LanguageTool, you can also rely on [this solution](https://stackoverflow.com/a/46306176/5875021) to install the older version and keep it from updating to the unsupported one. That way you don't need to bother with MacPorts, textidote, or install the LanguageTool manually, and the old configuration of `vim-LanguageTool` (e.g. `let g:languagetool_cmd='languagetool'`) from before this issue started to happen should work.

**Comment by Toreno96** [comment]
> (I actually don't have a "core" folder, in /opt/homebrew/core, so I had to adjust it a little bit, but still this is really cool)

@michaelfresco Yeah, it requires an additional step on the current version of Homebrew: https://stackoverflow.com/questions/39187812/how-to-install-older-versions-using-homebrew/46306176#comment137061519_46306176

**Comment by FausztBenedek** [comment]
I assembled a little hack for my neovim config that downloads the still compatible version into `~/.local/share/nvom/language-tool-old/LanguageTool-5.2/`, if it is not there yet, and sets the global variable accordingly. It does nothing if it is already downloaded.
```lua
    -- The plugin is incompatible with newer versions of languagetool. This downloads the older archived version:
    -- See issue https://github.com/dpelle/vim-LanguageTool/issues/33
    local Job = require("plenary.job")
    local url = "https://languagetool.org/download/archive/LanguageTool-5.2.zip"
    local language_tool_directory = vim.fn.expand('~') .. "/.local/share/nvim/language-tool-old"
    local zipfile = language_tool_directory .. "/LanguageTool-5.2.zip"
    vim.g.languagetool_jar = language_tool_directory .. "/LanguageTool-5.2/languagetool-commandline.jar"

    if directory_exists(language_tool_directory) then
      return
    end
    -- Download the file
    print("Start downloading LanguageTool")
    Job:new({
      command = "wget",
      args = { "-P", language_tool_directory, url },
      on_stdout = function(_, data)
        print(data)
      end,
      on_exit = function(j, return_val)
        print("LanguageTool downloaded")
        if return_val ~= 0 then
          print("Error unzipping file: " .. table.concat(j:stderr_result(), "\n"))
          return
        end
        Job:new({
          command = "unzip",
          args = { "-o", zipfile, "-d", language_tool_directory },
          on_exit = function(j, return_val)
            if return_val ~= 0 then
              print("Error unzipping file: " .. table.concat(j:stderr_result(), "\n"))
              return
            end
            print("LanguageTool downloaded and unzipped successfully!")
          end,
        }):start()
      end,
    }):start()
```

**Comment by tarikgraba** [comment]
If you are looking for old vesion (<6) they are still online https://languagetool.org/download/archive/

**Comment by bondhugula** [comment]
I can confirm that it still works with LanguageTool 5.9, which is available here: https://languagetool.org/download/archive/LanguageTool-5.9.zip
